server/lib/nl/nl_data_spec_next.py

```diff
@@ -449,6 +498,30 @@ def _places_from_context(uttr: Utterance) -> List[Place]:
   return ans
 
 
+def _places_for_comparison_from_context(uttr: Utterance) -> List[List[Place]]:
+  ans = []
+  first_uttr_count = 0
+  first = uttr
+  while (first and first_uttr_count < CTX_LOOKBACK_LIMIT + 1):
+    if len(first.places) == 1:
+      # Found only one place.  Try to find related place by looping up
+      # the utterance chain.
+      second = first.prev_utterance
+      second_uttr_count = first_uttr_count + 1
+      while (second and second_uttr_count < CTX_LOOKBACK_LIMIT + 1):
+        # For the first place, add all combinations of the second.
+        if second.places:
+          ans.append(second.places + first.places)
+        second = second.prev_utterance
+        second_uttr_count = second_uttr_count + 1
+    elif len(first.places) > 1:
+      # Found multiple places in a single utterance.  Use this.
+      ans.append(first.places)
```

Comment:
do you want to break here?

Reply:
We are returning multiple lists so the caller will try each related-places-list in order until successful.  So we don't want to break here?

Reply:
Can you add some comments or example for the choice here. For example for a query sequence of "tell me about san jose", "how is auto theft in fremont", "compare with palo alto". Will. the last one use all the places above (fremont, san jose)?

Reply:
Clarified, please take a look.